src/data/setoid.lean

```diff
@@ -0,0 +1,603 @@
+/-
+Copyright (c) 2019 Amelia Livingston. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amelia Livingston, Bryan Gin-ge Chen
+-/
+
+import data.quot data.set.lattice data.fintype order.order_iso order.galois_connection
+
+/-!
+# Equivalence relations
+
+The first section of the file defines the complete lattice of equivalence relations 
+on a type, results about the inductively defined equivalence closure of a binary relation, 
+and the analogues of some isomorphism theorems for quotients of arbitrary types.
+
+The second section comprises properties of equivalence relations viewed as partitions, and 
+a function "bell" that computes Bell numbers - the numbers of possible partitions of a finite 
+set - along with some preliminary results about "bell".
+
+## Implementation notes
+
+The function "rel" and lemmas ending in ' make it easier to talk about different 
+equivalence relations on the same type.
+
+The complete lattice instance for setoids could be defined by lifting the Galois
+insertion of equivalence relations on α into binary relations on α, but the resulting
+functions are less easy to use. 
+
+Partitions are not defined as a separate structure here; it is encouraged that users
+reason about them using the existing "setoid" and its infrastructure.
+
+Eventually there should be a proof here that Bell numbers are the number of possible
+partitions of a finite set.
+
+## References
+
+* https://en.wikipedia.org/wiki/Bell_triangle
+
+## Tags
+
+setoid, equivalence, iseqv, relation, equivalence relation, partition, equivalence 
+class, Bell number, Bell triangle
+-/
+
+variables {α : Type*} {β : Type*}
+
+open lattice 
+
+namespace order_iso
+
+variables [preorder α] [preorder β] (oi : @order_iso α β (≤) (≤))
+
+-- order.order_iso doesn't import order.galois_connection, and vice versa.
```

Comment:
I think it would make sense for `order.galois_connection` to import `order.order_iso`, and move these constructions to `order.galois_connection`.